internal/locCluster.ecl

```diff
@@ -0,0 +1,314 @@
+
+    #include<iostream>
+    #include<bits/stdc++.h>
+
+    using namespace std;
+
+    string distanceFunc = "euclidian";
+    vector<double> dist_params;
+    //data structure to represent the data points in a dataset for disjoint sets.
+    struct node{
+        uint16_t wi;
+        uint64_t id;
+        uint64_t nodeId;
+        vector<float> fields;
+        bool ifLocal = false;
+        bool ifCore = false;
+        node *parent = NULL;
+        bool isModified = false;
+        bool isVisited = false;
+    };
+    // distance = (sum((a - b)^2))^(1/2)
+    double euclidian(vector<float> a, vector<float> b){
+        double sum=0;
+        for(int i=0; i<a.size(); ++i){
+            sum += (a[i] - b[i])*(a[i] - b[i]);
+        }
+        return sqrt(sum);
+    }
+    // distance = 2*r*sin^-1((sin^2((a-b)/2)+cos(a)*cos(b)*sin^2((x-y)/2))
+    // applicable only when size of columns is 2.
+    double haversine(vector<float> a, vector<float> b){
+        int M=a.size();
+
+        if(M!=2){
+            return 0;
+        }
+
+        double lat1=a[0];
+        double lat2=b[0];
+        double lon1=a[1];
+        double lon2=b[1];
+
+        double sin_0 = sin(0.5 * (lat1 - lat2));
+        double sin_1 = sin(0.5 * (lon1 - lon2));
+
+        return (sin_0 * sin_0 + cos(lat1) * cos(lat2) * sin_1 * sin_1);
+    }
+
+    // distance = sum(|a - b|)
+    double manhattan(vector<float> a, vector<float> b){
+        double ans=0;
+        int M=a.size();
+        for(int i=0;i<M;i++)
+        ans=ans+(abs((a[i])-(b[i])));
+
+        return ans;
+    }
+    // distance = sum(|a - b|^p)^(1/p)
+    double minkowski(vector<float> a, vector<float> b, int p){
+        int m=a.size();
+        double ans=0;
+        for(int i=0;i<m;i++){
+                ans+=pow(abs(a[i]-b[i]),p);
+        }
+
+        return pow(ans,(double)1/p);
+    }
+
+    // distance = ((A.B)/(||A||.||B||))
+    double cosine(vector<float> a, vector<float> b){
+        double ans=0, a1=0,a2=0;
+        int m=a.size();
+        for(int i=0;i<m;i++){
+                ans+=a[i]*b[i];
+                a1=a1+pow(a[i],2);
+                a2=a2+pow(b[i],2);
+        }
+        return ans/(sqrt(a1)*sqrt(a2));
+    }
+    // distance = max(|x - y|)
+    double chebyshev(vector<float> a, vector<float> b){
+        int m=a.size();
+        float ans=0;
+        for(int i=0;i<m;i++){
+                ans=max(abs(a[i]-b[i]),ans);
+        }
+        return ans;
+    }
+    /*
+      getNeighbors Function to get the nearest neighbors within eps distance.
+      input: Vector ds, node p, distance eps
+      output: A vecor of nearest neighbors
+    */
+    vector<node*> getNeighbors(vector<node*> ds, node *p, double eps){
+        vector<node*> ret;
+        for(uint64_t i=0; i<ds.size(); ++i){
+
+            if(p->wi != ds[i]->wi) continue;
+
+            double dist;
+
+            if(distanceFunc.compare("cosine")==0)
```

Comment:
This will not be constant folded and the overhead may be noticable. It might be better to use a different mechanism to select which distance function to use.  (E.g. a pointer to function or enumeration.)

Reply:
If it's okay, I prefer to updating it in the next stage.

Reply:
Yes, I think it would be better to updated it separately.